Cannot load JAR packaged `application.properties` file with Spring framework when switching to Java 11

Dear Azure Functions Java team,

We're trying to upgrade our Java based Azure functions to Java 11, but the Spring framework cannot load JAR packaged `application.properties` file when switching our functions to Java 11.

Java and Maven versions:

```
Apache Maven 3.8.1 (05c21c65bdfed0f71a2f2ada8b84da59348c4c5d)
Java version: 11.0.15, vendor: Azul Systems, Inc., runtime: /Users/cbismuth/Library/Java/JavaVirtualMachines/azul-11.0.15/Contents/Home
```

We've already upgraded to the latest Azure runtime v4, and here are the latest Maven dependency versions we use:

```
<azure-functions-maven-plugin.version>1.18.0</azure-functions-maven-plugin.version>
<azure-functions-java-library.version>2.0.1</azure-functions-java-library.version>
```

We can reproduce the issue on our laptops using `mvn -U azure-functions:run` Maven command line.

Using Java 8 everything works fine, but when switching to Java 11 the Spring application context failed to load in the function with the error below.

(A simple `main`-like application just loading the Spring application context works fine with Java 11.)

Error:

`Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [fr.tf1.data.commons.config.SpringConfig]; nested exception is java.io.FileNotFoundException: class path resource [application.properties] cannot be opened because it does not exist`

Stack trace:

```
Azure Functions Core Tools
Core Tools Version:       4.0.4590 Commit hash: N/A  (64-bit)
Function Runtime Version: 4.5.2.18383


Functions:

        git: [GET] http://localhost:7071/api/git

        graphRecommendProgramIds: [GET] http://localhost:7071/api/v1/programs/{userId}

        graphRecommendVideoIds: [GET] http://localhost:7071/api/v1/videos/{userId}

        health: [GET] http://localhost:7071/api/health

        jvm: [GET] http://localhost:7071/api/jvm

For detailed output, run func with --verbose flag.
[2022-06-20T14:39:26.004Z] Worker process started and initialized.
[2022-06-20T14:39:26.078Z] [WARNING] {JavaFunctionBroker.addJavaAnnotationLibrary}: Customer is not providing java annotation files, they may not use the latest version of java library and maven plugin.
[2022-06-20T14:39:26.079Z] [WARNING] {JavaFunctionBroker.addJavaAnnotationLibrary}: Customer is not providing java annotation files, they may not use the latest version of java library and maven plugin.
[2022-06-20T14:39:26.079Z] [WARNING] {JavaFunctionBroker.addJavaAnnotationLibrary}: Customer is not providing java annotation files, they may not use the latest version of java library and maven plugin.
[2022-06-20T14:39:26.079Z] [WARNING] {JavaFunctionBroker.addJavaAnnotationLibrary}: Customer is not providing java annotation files, they may not use the latest version of java library and maven plugin.
[2022-06-20T14:39:26.079Z] [WARNING] {JavaFunctionBroker.addJavaAnnotationLibrary}: Customer is not providing java annotation files, they may not use the latest version of java library and maven plugin.
[2022-06-20T14:39:26.562Z] juin 20, 2022 4:39:26 PM org.springframework.context.annotation.AnnotationConfigApplicationContext refresh
[2022-06-20T14:39:26.562Z] AVERTISSEMENT: Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [fr.tf1.data.commons.config.SpringConfig]; nested exception is java.io.FileNotFoundException: class path resource [application.properties] cannot be opened because it does not exist
[2022-06-20T14:39:26.572Z] 16:39:26.570 [pool-2-thread-3] ERROR f.t.d.c.config.SpringBeanProvider - Failed to parse configuration class [fr.tf1.data.commons.config.SpringConfig]; nested exception is java.io.FileNotFoundException: class path resource [application.properties] cannot be opened because it does not exist org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [fr.tf1.data.commons.config.SpringConfig]; nested exception is java.io.FileNotFoundException: class path resource [application.properties] cannot be opened because it does not exist
    at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:189)
    at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:331)
    at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:247)
    at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:311)
    at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:112)
    at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:746)
    at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:564)
    at fr.tf1.data.commons.config.SpringBeanProvider.<init>(SpringBeanProvider.java:96)
    at fr.tf1.data.commons.config.SpringBeanProvider.initialize(SpringBeanProvider.java:47)
    at fr.tf1.data.commons.config.SpringBeanProvider.getInstance(SpringBeanProvider.java:30)
    at fr.tf1.data.commons.config.SpringBeanProvider.getBean(SpringBeanProvider.java:57)
    at fr.tf1.data.recommendation.graph.function.AbstractRecommendationFunction.<clinit>(AbstractRecommendationFunction.java:31)
    at java.base/java.lang.Class.forName0(Native Method)
    at java.base/java.lang.Class.forName(Class.java:398)
    at com.microsoft.azure.functions.worker.broker.EnhancedJavaMethodExecutorImpl.getContainingClass(EnhancedJavaMethodExecutorImpl.java:63)
    at com.microsoft.azure.functions.worker.broker.EnhancedJavaMethodExecutorImpl.<init>(EnhancedJavaMethodExecutorImpl.java:22)
    at com.microsoft.azure.functions.worker.broker.FactoryJavaMethodExecutor.getJavaMethodExecutor(FactoryJavaMethodExecutor.java:20)
    at com.microsoft.azure.functions.worker.broker.JavaFunctionBroker.loadMethod(JavaFunctionBroker.java:38)
    at com.microsoft.azure.functions.worker.handler.FunctionLoadRequestHandler.execute(FunctionLoadRequestHandler.java:27)
    at com.microsoft.azure.functions.worker.handler.FunctionLoadRequestHandler.execute(FunctionLoadRequestHandler.java:9)
    at com.microsoft.azure.functions.worker.handler.MessageHandler.handle(MessageHandler.java:45)
    at com.microsoft.azure.functions.worker.JavaWorkerClient$StreamingMessagePeer.lambda$onNext$0(JavaWorkerClient.java:92)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.io.FileNotFoundException: class path resource [application.properties] cannot be opened because it does not exist
    at org.springframework.core.io.ClassPathResource.getInputStream(ClassPathResource.java:199)
    at org.springframework.core.io.support.EncodedResource.getInputStream(EncodedResource.java:159)
    at org.springframework.core.io.support.PropertiesLoaderUtils.fillProperties(PropertiesLoaderUtils.java:110)
    at org.springframework.core.io.support.PropertiesLoaderUtils.fillProperties(PropertiesLoaderUtils.java:81)
    at org.springframework.core.io.support.PropertiesLoaderUtils.loadProperties(PropertiesLoaderUtils.java:67)
    at org.springframework.core.io.support.ResourcePropertySource.<init>(ResourcePropertySource.java:67)
    at org.springframework.core.io.support.DefaultPropertySourceFactory.createPropertySource(DefaultPropertySourceFactory.java:37)
    at org.springframework.context.annotation.ConfigurationClassParser.processPropertySource(ConfigurationClassParser.java:463)
    at org.springframework.context.annotation.ConfigurationClassParser.doProcessConfigurationClass(ConfigurationClassParser.java:280)
    at org.springframework.context.annotation.ConfigurationClassParser.processConfigurationClass(ConfigurationClassParser.java:250)
    at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:207)
    at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:175)
... 26 common frames omitted

[2022-06-20T14:39:26.606Z] 16:39:26.572 [pool-2-thread-3] ERROR f.t.d.c.config.SpringBeanProvider - org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [fr.tf1.data.commons.config.SpringConfig]; nested exception is java.io.FileNotFoundException: class path resource [application.properties] cannot be opened because it does not exist java.lang.RuntimeException: org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [fr.tf1.data.commons.config.SpringConfig]; nested exception is java.io.FileNotFoundException: class path resource [application.properties] cannot be opened because it does not exist
    at fr.tf1.data.commons.config.SpringBeanProvider.<init>(SpringBeanProvider.java:104)
    at fr.tf1.data.commons.config.SpringBeanProvider.initialize(SpringBeanProvider.java:47)
    at fr.tf1.data.commons.config.SpringBeanProvider.getInstance(SpringBeanProvider.java:30)
    at fr.tf1.data.commons.config.SpringBeanProvider.getBean(SpringBeanProvider.java:57)
    at fr.tf1.data.recommendation.graph.function.AbstractRecommendationFunction.<clinit>(AbstractRecommendationFunction.java:31)
    at java.base/java.lang.Class.forName0(Native Method)
    at java.base/java.lang.Class.forName(Class.java:398)
    at com.microsoft.azure.functions.worker.broker.EnhancedJavaMethodExecutorImpl.getContainingClass(EnhancedJavaMethodExecutorImpl.java:63)
    at com.microsoft.azure.functions.worker.broker.EnhancedJavaMethodExecutorImpl.<init>(EnhancedJavaMethodExecutorImpl.java:22)
    at com.microsoft.azure.functions.worker.broker.FactoryJavaMethodExecutor.getJavaMethodExecutor(FactoryJavaMethodExecutor.java:20)
    at com.microsoft.azure.functions.worker.broker.JavaFunctionBroker.loadMethod(JavaFunctionBroker.java:38)
    at com.microsoft.azure.functions.worker.handler.FunctionLoadRequestHandler.execute(FunctionLoadRequestHandler.java:27)
    at com.microsoft.azure.functions.worker.handler.FunctionLoadRequestHandler.execute(FunctionLoadRequestHandler.java:9)
    at com.microsoft.azure.functions.worker.handler.MessageHandler.handle(MessageHandler.java:45)
    at com.microsoft.azure.functions.worker.JavaWorkerClient$StreamingMessagePeer.lambda$onNext$0(JavaWorkerClient.java:92)
    at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
    at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
    at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
    at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
    at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.springframework.beans.factory.BeanDefinitionStoreException: Failed to parse configuration class [fr.tf1.data.commons.config.SpringConfig]; nested exception is java.io.FileNotFoundException: class path resource [application.properties] cannot be opened because it does not exist
    at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:189)
    at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:331)
    at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:247)
    at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:311)
    at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:112)
    at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:746)
    at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:564)
    at fr.tf1.data.commons.config.SpringBeanProvider.<init>(SpringBeanProvider.java:96)
... 19 common frames omitted
Caused by: java.io.FileNotFoundException: class path resource [application.properties] cannot be opened because it does not exist
    at org.springframework.core.io.ClassPathResource.getInputStream(ClassPathResource.java:199)
    at org.springframework.core.io.support.EncodedResource.getInputStream(EncodedResource.java:159)
    at org.springframework.core.io.support.PropertiesLoaderUtils.fillProperties(PropertiesLoaderUtils.java:110)
    at org.springframework.core.io.support.PropertiesLoaderUtils.fillProperties(PropertiesLoaderUtils.java:81)
    at org.springframework.core.io.support.PropertiesLoaderUtils.loadProperties(PropertiesLoaderUtils.java:67)
    at org.springframework.core.io.support.ResourcePropertySource.<init>(ResourcePropertySource.java:67)
    at org.springframework.core.io.support.DefaultPropertySourceFactory.createPropertySource(DefaultPropertySourceFactory.java:37)
    at org.springframework.context.annotation.ConfigurationClassParser.processPropertySource(ConfigurationClassParser.java:463)
    at org.springframework.context.annotation.ConfigurationClassParser.doProcessConfigurationClass(ConfigurationClassParser.java:280)
    at org.springframework.context.annotation.ConfigurationClassParser.processConfigurationClass(ConfigurationClassParser.java:250)
    at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:207)
    at org.springframework.context.annotation.ConfigurationClassParser.parse(ConfigurationClassParser.java:175)
... 26 common frames omitted

[2022-06-20T14:39:30.098Z] Host lock lease acquired by instance ID '000000000000000000000000743AEE76'.
```

Spring application context:

```java
@Configuration
@EnableScheduling
@EnableTransactionManagement
@ComponentScan("fr.tf1.data")
@PropertySource("classpath:application.properties")
public class SpringConfig {

    private static final Logger LOGGER = LoggerFactory.getLogger(SpringConfig.class);

    @Bean
    public static PropertySourcesPlaceholderConfigurer propertyPlaceholderConfigurer() {
        return new PropertySourcesPlaceholderConfigurer();
    }

    @Bean
    public ConversionService conversionService() {
        return new DefaultConversionService();
    }

    @Bean
    public PersistenceExceptionTranslationPostProcessor persistenceExceptionTranslationPostProcessor() {
        return new PersistenceExceptionTranslationPostProcessor();
    }
}
```

Spring application loading code:

```java
    private SpringBeanProvider(final String... args) {
        final Stopwatch stopwatch = Stopwatch.createStarted();

        try {
            applicationContext = new AnnotationConfigApplicationContext();

            if (args.length != 0) {
                applicationContext.getEnvironment()
                                  .getPropertySources()
                                  .addFirst(new SimpleCommandLinePropertySource(args));
            }

            applicationContext.register(SpringConfig.class);

            applicationContext.refresh();

            Runtime.getRuntime().addShutdownHook(new Thread(applicationContext::close));

            LOGGER.info("Spring application context successfully started in [{}]", DateHelper.toPrintablePeriod(stopwatch));
        } catch (final Throwable e) {
            LOGGER.error(e.getMessage(), e);

            throw new RuntimeException(e);
        }
    }
```

--- a/src/main/java/com/microsoft/azure/functions/worker/broker/EnhancedJavaMethodExecutorImpl.java
+++ b/src/main/java/com/microsoft/azure/functions/worker/broker/EnhancedJavaMethodExecutorImpl.java
@@ -60,7 +60,7 @@ public class EnhancedJavaMethodExecutorImpl implements JavaMethodExecutor {
     }
 
     private Class<?> getContainingClass(String className) throws ClassNotFoundException {
-        return Class.forName(className, true, this.classLoader);
+        return Class.forName(className, false, this.classLoader);
     }
 
     private final Class<?> containingClass;

